src/js_native_api_v8.cc

```diff
@@ -1662,6 +1662,28 @@ napi_status napi_create_symbol(napi_env env,
   return napi_clear_last_error(env);
 }
 
+napi_status napi_create_symbol_for(napi_env env,
+                                   napi_value description,
+                                   napi_value* result) {
+  CHECK_ENV(env);
+  CHECK_ARG(env, result);
+
+  v8::Isolate* isolate = env->isolate;
+
+  v8::Local<v8::Value> desc;
+  if (description == nullptr) {
+    CHECK_NEW_FROM_UTF8(env, desc, "undefined");
```

Comment:
Should this really be allowed in the first place?

Reply:
Yes, Symbol.for() can be called without any arguments, so the native variation too should provide a way of doing that. napi_create_symbol() too implements calls to Symbol() without arguments by treating null descriptions separately.

Reply:
Well … but this is the case in which nullptr is *explicitly* passed as an argument. I can’t imagine a situation in which this would not be a bug.

> napi_create_symbol() too implements calls to Symbol() without arguments by treating null descriptions separately.

But that has meaningful semantics – `Symbol()` and `Symbol.for()` are very different things. The first one a programmer might very well write on purpose, because the argument is a description and is only useful for diagnostic purposes; the second one is almost certainly a bug or typo, because the argument is *not* just used for diagnostic purposes but instead an actual global identifier.

Reply:
It might be a source of bugs but I still believe that we should allow it as long as the language allows it. Should we create a tc39 proposal to throw an error if Symbol.for() is called without any arguments? I'd also be interested in knowing how Symbol() can be used for diagnostic purposes.

Reply:
> It might be a source of bugs but I still believe that we should allow it as long as the language allows it. 

Frankly, this is inconsistent and would set a bad precedent.

Your current implementation allows either `nullptr` or a string. The language allows *any* value (including no value, which corresponds to passing `undefined` as an argument) and *converts* it to string.

If the goal is to be consistent with the language (to the degree where we willingly accept bugs that come from it), then you should also accept *any* value and convert to string according to language rules, but still *not* `nullptr`, because passing `nullptr` is not the same as passing no argument – it’s explicitly passing a type of value that doesn’t even exist in JS.

> Should we create a tc39 proposal to throw an error if Symbol.for() is called without any arguments?

I don’t think this kind of breaking change would qualify as a viable proposal, and it would probably also be inconsistent with the language in general, which leans towards doing type conversions rather than type checking.

> I'd also be interested in knowing how Symbol() can be used for diagnostic purposes.

For example, in Node.js core, we make the Symbol description match the variable name, e.g. `const kHeaders = Symbol('kHeaders');`, so that if the Symbol is shown in diagnostic output, we can easily find where it originated from.

Reply:
> If the goal is to be consistent with the language (to the degree where we willingly accept bugs that come from it), then you should also accept any value and convert to string according to language rules;

Hmm, if we do that for the symbol creation functions, we'll have to do that for the rest of the API and that would break the ABI, so it's not worth it. I'll change the implementation to reject null descriptions then but I'm still not sure about the use of accepting nullptr descriptions in napi_create_symbol().

> but still not nullptr, because passing nullptr is not the same as passing no argument – it’s explicitly passing a type of value that doesn’t even exist in JS.

I treated null as 0 params because napi_create_symbol() was doing so.

> For example, in Node.js core, we make the Symbol description match the variable name, e.g. const kHeaders = Symbol('kHeaders');, so that if the Symbol is shown in diagnostic output, we can easily find where it originated from.

I meant using Symbol() without any arguments because you were comparing that to calling Symbol.for() without any arguments. :)

Reply:
> > If the goal is to be consistent with the language (to the degree where we willingly accept bugs that come from it), then you should also accept any value and convert to string according to language rules;
> 
> Hmm, if we do that for the symbol creation functions, we'll have to do that for the rest of the API and that would break the ABI, so it's not worth it. I'll change the implementation to reject null descriptions then 

To be clear, that’s also what I would prefer, over accepting any value in general :) Esp. because many of the languages that use N-API (C, C++, Rust) are typesafe anyway.

Reply:
@addaleax I've updated the PR to disallow null descriptions, PTAL